modules/assim.sequential/R/Analysis_sda_block.R

```diff
@@ -0,0 +1,549 @@
+    return(0)
+  }
+  
+  #update MCMC args.
+  block.list.all[[t]] <- block.list.all[[t]] %>% 
+    purrr::map(function(l){
+      l$MCMC <- MCMC.args
+      l
+    })
+  
+  #parallel for loop over each block.
+  PEcAn.logger::logger.info(paste0("Running MCMC ", "for ", length(block.list.all[[t]]), " blocks"))
+  if ("try-error" %in% class(try(block.list.all[[t]] <- furrr::future_map(block.list.all[[t]], MCMC_block_function, .progress = T)))) {
+    PEcAn.logger::logger.error("Something wrong within the MCMC_block_function function.")
+    return(0)
+  }
+  PEcAn.logger::logger.info("Completed!")
+  
+  #convert from block lists to vector values.
+  if ("try-error" %in% class(try(V <- block.2.vector(block.list.all[[t]], X, H)))) {
+    PEcAn.logger::logger.error("Something wrong within the block.2.vector function.")
+    return(0)
+  }
+  
+  #return values
+  return(list(block.list.all = block.list.all,
+         mu.f = V$mu.f,
+         Pf = V$Pf,
+         mu.a = V$mu.a,
+         Pa = V$Pa,
+         Y = Y,
+         R = R))
+}
+
+##' @title build.block.xy
+##' @name  build.block.xy
+##' @author Dongchen Zhang
+##' 
+##' @param settings  pecan standard multi-site settings list.  
+##' @param block.list.all List contains nt empty sub-elements.
+##' @param X A matrix contains ensemble forecasts.
+##' @param obs.mean List of dataframe of observation means, named with observation datetime.
+##' @param obs.cov   List of covariance matrices of state variables , named with observation datetime.
+##' @param t time point.
+##' @details This function will add data and constants into each block that are needed for the MCMC sampling.
+##'  
+##' @description This function split long vector and covariance matrix into blocks corresponding to the localization.
+##' 
+##' @return It returns the `build.block.xy` object with data and constants filled in.
+build.block.xy <- function(settings, block.list.all, X, obs.mean, obs.cov, t) {
+  #set q.type from settings.
+  if (settings$state.data.assimilation$q.type == "vector") {
+    q.type <- 1
+  } else if (settings$state.data.assimilation$q.type == "wishart") {
+    q.type <- 2
+  }
+  
+  #grab basic arguments based on X.
+  site.ids <- unique(attributes(X)$Site)
+  var.names <- unique(attributes(X)$dimnames[[2]])
+  mu.f <- colMeans(X)
+  Pf <- stats::cov(X)
+  diag(Pf)[which(diag(Pf)==0)] <- min(diag(Pf)[which(diag(Pf) != 0)])/5 #fixing det(Pf)==0
+  
+  #distance calculations and localization
+  if (!is.null(settings$state.data.assimilation$Localization.FUN)) {
+    Localization.FUN <- get(settings$state.data.assimilation$Localization.FUN)
+    site.locs <- settings$run %>% 
+      purrr::map('site') %>% 
+      purrr::map_dfr(~c(.x[['lon']],.x[['lat']]) %>% as.numeric)%>% 
+      t %>%
+      `colnames<-`(c("Lon","Lat")) %>%
+      `rownames<-`(site.ids)
+    #Finding the distance between the sites
+    dis.matrix <- sp::spDists(site.locs, longlat = TRUE)
```

Comment:
below it appears that dis.matrix is required for the block calculation. You probably want to move site.locs and dis.matrix calculations to before this `if` since, as written, the existence of settings$state.data.assimilation$Localization.FUN is optional.

Reply:
Fixed, thanks for catching that!